MembershipUpdateTest.connectionsToRemovedMember_shouldBeClosed failure

*maintenance*

Failed on PR builder: https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/new-lab-fast-pr/15425/testReport/junit/com.hazelcast.internal.cluster.impl/MembershipUpdateTest/connectionsToRemovedMember_shouldBeClosed/

Stacktrace:
```
java.lang.AssertionError: expected null, but was:<com.hazelcast.nio.tcp.DroppingConnection@2fea7129>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotNull(Assert.java:755)
	at org.junit.Assert.assertNull(Assert.java:737)
	at org.junit.Assert.assertNull(Assert.java:747)
	at com.hazelcast.internal.cluster.impl.MembershipUpdateTest.connectionsToRemovedMember_shouldBeClosed(MembershipUpdateTest.java:797)
```

Lots of exceptions on standard output:
```
07:28:09,611  WARN |noOperationExecuted_beforePreJoinOpIsDone| - [MembershipUpdateTest$ExecutionTrackerOp] hz._hzInstance_81_dev.generic-operation.thread-24 - [127.0.0.1]:5081 [dev] [3.10.1-SNAPSHOT] com.hazelcast.spi.exception.CallerNotMemberException: Not Member! this: [127.0.0.1]:5081, caller: [127.0.0.1]:5079, partitionId: -1, operation: com.hazelcast.internal.cluster.impl.MembershipUpdateTest$ExecutionTrackerOp, service: pre-join-service
```

```
Hiccups measured while running test 'connectionsToRemovedMember_shouldBeClosed(com.hazelcast.internal.cluster.impl.MembershipUpdateTest):'
07:27:50, accumulated pauses: 2121 ms, max pause: 1132 ms, pauses over 1000 ms: 1
07:27:55, accumulated pauses: 1048 ms, max pause: 326 ms, pauses over 1000 ms: 0
07:28:00, accumulated pauses: 1439 ms, max pause: 236 ms, pauses over 1000 ms: 0
07:28:05, accumulated pauses: 671 ms, max pause: 69 ms, pauses over 1000 ms: 0
07:28:10, accumulated pauses: 1392 ms, max pause: 654 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/MembershipUpdateTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/MembershipUpdateTest.java
@@ -29,9 +29,10 @@ import com.hazelcast.instance.NodeExtension;
 import com.hazelcast.internal.cluster.MemberInfo;
 import com.hazelcast.internal.cluster.impl.operations.MembersUpdateOp;
 import com.hazelcast.nio.Address;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.nio.ConnectionListener;
 import com.hazelcast.nio.ConnectionManager;
 import com.hazelcast.nio.ObjectDataInput;
-import com.hazelcast.nio.tcp.FirewallingConnectionManager;
 import com.hazelcast.spi.Operation;
 import com.hazelcast.spi.OperationService;
 import com.hazelcast.spi.PostJoinAwareService;
@@ -760,13 +761,20 @@ public class MembershipUpdateTest extends HazelcastTestSupport {
 
         assertClusterSizeEventually(3, hz1, hz2, hz3);
 
-        Address target = getAddress(hz2);
+        final Address target = getAddress(hz2);
         hz2.getLifecycleService().terminate();
 
         assertClusterSizeEventually(2, hz1, hz3);
 
         assertNull(getConnectionManager(hz1).getConnection(target));
-        assertNull(getConnectionManager(hz3).getConnection(target));
+
+        final ConnectionManager cm3 = getConnectionManager(hz3);
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertNull(cm3.getConnection(target));
+            }
+        });
     }
 
     @Test
@@ -781,24 +789,19 @@ public class MembershipUpdateTest extends HazelcastTestSupport {
 
         assertClusterSizeEventually(3, hz1, hz2, hz3);
 
-        Address target = getAddress(hz2);
+        final Address target = getAddress(hz3);
 
-        FirewallingConnectionManager cm2 = getFireWallingConnectionManager(hz2);
-        cm2.blockNewConnection(getAddress(hz1));
-        cm2.blockNewConnection(getAddress(hz3));
+        ConnectionRemovedListener connListener1 = new ConnectionRemovedListener(target);
+        getConnectionManager(hz1).addConnectionListener(connListener1);
 
-        FirewallingConnectionManager cm3 = getFireWallingConnectionManager(hz3);
-        cm3.blockNewConnection(getAddress((hz2)));
+        ConnectionRemovedListener connListener2 = new ConnectionRemovedListener(target);
+        getConnectionManager(hz2).addConnectionListener(connListener2);
 
-        dropOperationsBetween(hz2, hz1, F_ID, singletonList(HEARTBEAT));
-        assertClusterSizeEventually(2, hz1, hz3);
+        dropOperationsBetween(hz3, hz1, F_ID, singletonList(HEARTBEAT));
+        assertClusterSizeEventually(2, hz1, hz2);
 
-        assertNull(getConnectionManager(hz1).getConnection(target));
-        assertNull(getConnectionManager(hz3).getConnection(target));
-    }
-
-    private static FirewallingConnectionManager getFireWallingConnectionManager(HazelcastInstance hz) {
-        return (FirewallingConnectionManager) getConnectionManager(hz);
+        connListener1.assertConnectionRemoved();
+        connListener2.assertConnectionRemoved();
     }
 
     private Config getConfigWithService(Object service, String serviceName) {
@@ -969,6 +972,30 @@ public class MembershipUpdateTest extends HazelcastTestSupport {
         @Override
         protected void readInternal(ObjectDataInput in) throws IOException {
             throw new RuntimeException("This operation always fails during deserialization");
+        }
+    }
+
+    private static class ConnectionRemovedListener implements ConnectionListener {
+        private final Address endpoint;
+        private final CountDownLatch latch = new CountDownLatch(1);
+
+        ConnectionRemovedListener(Address endpoint) {
+            this.endpoint = endpoint;
+        }
+
+        @Override
+        public void connectionAdded(Connection connection) {
+        }
+
+        @Override
+        public void connectionRemoved(Connection connection) {
+            if (endpoint.equals(connection.getEndPoint())) {
+                latch.countDown();
+            }
+        }
+
+        void assertConnectionRemoved() {
+            assertOpenEventually(latch);
         }
     }
 }

